### ISSUE TITLE
Python crashes when importing libbgs

### ISSUE THREAD

**Issue by Baozhang-Ren** [andrewssobral/bgslibrary#104]
Python crashes every time I import libbgs on mac.

Don't know what is the problem.

Please help!

`An error ocurred while starting the kernel
PYTHON TYPE CONVERTERS exported
Failed to import the site module
Traceback (most recent call last):
File "/Users/baozhang/anaconda3/lib/python3.6/site.py", line 541, in 
main()
File "/Users/baozhang/anaconda3/lib/python3.6/site.py", line 522, in main
known_paths = addusersitepackages(known_paths)
File "/Users/baozhang/anaconda3/lib/python3.6/site.py", line 282, in addusersitepackages
user_site = getusersitepackages()
File "/Users/baozhang/anaconda3/lib/python3.6/site.py", line 258, in getusersitepackages
user_base = getuserbase() # this will also set USER_BASE
File "/Users/baozhang/anaconda3/lib/python3.6/site.py", line 248, in getuserbase
USER_BASE = get_config_var('userbase')
File "/Users/baozhang/anaconda3/lib/python3.6/sysconfig.py", line 659, in get_config_var
return get_config_vars().get(name)
File "/Users/baozhang/anaconda3/lib/python3.6/sysconfig.py", line 633, in get_config_vars
conda_customize_compiler(_CONFIG_VARS)
File "/Users/baozhang/anaconda3/lib/python3.6/sysconfig.py", line 528, in conda_customize_compiler
from _osx_support import _find_executable
File "/Users/baozhang/anaconda3/lib/python3.6/_osx_support.py", line 4, in 
import re
File "/Users/baozhang/anaconda3/lib/python3.6/re.py", line 122, in 
import enum
File "/Users/baozhang/anaconda3/lib/python3.6/enum.py", line 2, in 
from types import MappingProxyType, DynamicClassAttribute
File "/Users/baozhang/anaconda3/lib/python3.6/types.py", line 171, in 
import functools as _functools
File "/Users/baozhang/anaconda3/lib/python3.6/functools.py", line 21, in 
from collections import namedtuple
File "/Users/baozhang/anaconda3/lib/python3.6/collections/__init__.py", line 29, in 
import heapq as _heapq
File "/Users/baozhang/anaconda3/lib/python3.6/heapq.py", line 587, in 
from _heapq import *
SystemError: initialization of _heapq did not return an extension module`

**Comment by andrewssobral** [comment]
Hi @Baozhang-Ren , thank you for reporting this issue. I am having the same problem, I am looking at this...

**Comment by andrewssobral** [comment]
Now I'm getting this error while building bgslibrary with python support:
```
[100%] Linking CXX shared library libbgs.dylib
/usr/local/Cellar/cmake/3.9.1/bin/cmake -E cmake_link_script CMakeFiles/libbgs.dir/link.txt --verbose=1
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/c++  -std=gnu++0x -O3 -DNDEBUG -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.13.sdk -dynamiclib -Wl,-headerpad_max_install_names  -o libbgs.dylib -install_name @rpath/libbgs.dylib CMakeFiles/libbgs.dir/package_bgs/AdaptiveBackgroundLearning.cpp.o CMakeFiles/libbgs.dir/package_bgs/AdaptiveSelectiveBackgroundLearning.cpp.o CMakeFiles/libbgs.dir/package_bgs/CodeBook.cpp.o CMakeFiles/libbgs.dir/package_bgs/DPAdaptiveMedian.cpp.o CMakeFiles/libbgs.dir/package_bgs/DPEigenbackground.cpp.o CMakeFiles/libbgs.dir/package_bgs/DPGrimsonGMM.cpp.o CMakeFiles/libbgs.dir/package_bgs/DPMean.cpp.o CMakeFiles/libbgs.dir/package_bgs/DPPratiMediod.cpp.o CMakeFiles/libbgs.dir/package_bgs/DPTexture.cpp.o CMakeFiles/libbgs.dir/package_bgs/DPWrenGA.cpp.o CMakeFiles/libbgs.dir/package_bgs/DPZivkovicAGMM.cpp.o CMakeFiles/libbgs.dir/package_bgs/FrameDifference.cpp.o CMakeFiles/libbgs.dir/package_bgs/FuzzyChoquetIntegral.cpp.o CMakeFiles/libbgs.dir/package_bgs/FuzzySugenoIntegral.cpp.o CMakeFiles/libbgs.dir/package_bgs/GMG.cpp.o CMakeFiles/libbgs.dir/package_bgs/IMBS/IMBS.cpp.o CMakeFiles/libbgs.dir/package_bgs/IndependentMultimodal.cpp.o CMakeFiles/libbgs.dir/package_bgs/KDE.cpp.o CMakeFiles/libbgs.dir/package_bgs/KDE/KernelTable.cpp.o CMakeFiles/libbgs.dir/package_bgs/KDE/NPBGSubtractor.cpp.o CMakeFiles/libbgs.dir/package_bgs/KDE/NPBGmodel.cpp.o CMakeFiles/libbgs.dir/package_bgs/KNN.cpp.o CMakeFiles/libbgs.dir/package_bgs/LBAdaptiveSOM.cpp.o CMakeFiles/libbgs.dir/package_bgs/LBFuzzyAdaptiveSOM.cpp.o CMakeFiles/libbgs.dir/package_bgs/LBFuzzyGaussian.cpp.o CMakeFiles/libbgs.dir/package_bgs/LBMixtureOfGaussians.cpp.o CMakeFiles/libbgs.dir/package_bgs/LBP_MRF.cpp.o CMakeFiles/libbgs.dir/package_bgs/LBP_MRF/MEDefs.cpp.o CMakeFiles/libbgs.dir/package_bgs/LBP_MRF/MEHistogram.cpp.o CMakeFiles/libbgs.dir/package_bgs/LBP_MRF/MEImage.cpp.o CMakeFiles/libbgs.dir/package_bgs/LBP_MRF/MotionDetection.cpp.o CMakeFiles/libbgs.dir/package_bgs/LBP_MRF/graph.cpp.o CMakeFiles/libbgs.dir/package_bgs/LBP_MRF/maxflow.cpp.o CMakeFiles/libbgs.dir/package_bgs/LBSP/BackgroundSubtractorLBSP.cpp.o CMakeFiles/libbgs.dir/package_bgs/LBSP/BackgroundSubtractorLBSP_.cpp.o CMakeFiles/libbgs.dir/package_bgs/LBSP/BackgroundSubtractorLOBSTER.cpp.o CMakeFiles/libbgs.dir/package_bgs/LBSP/BackgroundSubtractorPAWCS.cpp.o CMakeFiles/libbgs.dir/package_bgs/LBSP/BackgroundSubtractorSuBSENSE.cpp.o CMakeFiles/libbgs.dir/package_bgs/LBSP/LBSP.cpp.o CMakeFiles/libbgs.dir/package_bgs/LBSP/LBSP_.cpp.o CMakeFiles/libbgs.dir/package_bgs/LBSimpleGaussian.cpp.o CMakeFiles/libbgs.dir/package_bgs/LOBSTER.cpp.o CMakeFiles/libbgs.dir/package_bgs/MixtureOfGaussianV1.cpp.o CMakeFiles/libbgs.dir/package_bgs/MixtureOfGaussianV2.cpp.o CMakeFiles/libbgs.dir/package_bgs/MultiCue.cpp.o CMakeFiles/libbgs.dir/package_bgs/MultiLayer.cpp.o CMakeFiles/libbgs.dir/package_bgs/MultiLayer/BlobExtraction.cpp.o CMakeFiles/libbgs.dir/package_bgs/MultiLayer/BlobResult.cpp.o CMakeFiles/libbgs.dir/package_bgs/MultiLayer/CMultiLayerBGS.cpp.o CMakeFiles/libbgs.dir/package_bgs/MultiLayer/LocalBinaryPattern.cpp.o CMakeFiles/libbgs.dir/package_bgs/MultiLayer/blob.cpp.o CMakeFiles/libbgs.dir/package_bgs/PAWCS.cpp.o CMakeFiles/libbgs.dir/package_bgs/PBAS/PBAS.cpp.o CMakeFiles/libbgs.dir/package_bgs/PixelBasedAdaptiveSegmenter.cpp.o CMakeFiles/libbgs.dir/package_bgs/SigmaDelta.cpp.o CMakeFiles/libbgs.dir/package_bgs/SigmaDelta/sdLaMa091.cpp.o CMakeFiles/libbgs.dir/package_bgs/StaticFrameDifference.cpp.o CMakeFiles/libbgs.dir/package_bgs/SuBSENSE.cpp.o CMakeFiles/libbgs.dir/package_bgs/T2F/FuzzyUtils.cpp.o CMakeFiles/libbgs.dir/package_bgs/T2F/MRF.cpp.o CMakeFiles/libbgs.dir/package_bgs/T2F/T2FGMM.cpp.o CMakeFiles/libbgs.dir/package_bgs/T2F/T2FMRF.cpp.o CMakeFiles/libbgs.dir/package_bgs/T2FGMM_UM.cpp.o CMakeFiles/libbgs.dir/package_bgs/T2FGMM_UV.cpp.o CMakeFiles/libbgs.dir/package_bgs/T2FMRF_UM.cpp.o CMakeFiles/libbgs.dir/package_bgs/T2FMRF_UV.cpp.o CMakeFiles/libbgs.dir/package_bgs/TwoPoints.cpp.o CMakeFiles/libbgs.dir/package_bgs/TwoPoints/two_points.cpp.o CMakeFiles/libbgs.dir/package_bgs/ViBe.cpp.o CMakeFiles/libbgs.dir/package_bgs/ViBe/vibe-background-sequential.cpp.o CMakeFiles/libbgs.dir/package_bgs/VuMeter.cpp.o CMakeFiles/libbgs.dir/package_bgs/VuMeter/TBackground.cpp.o CMakeFiles/libbgs.dir/package_bgs/VuMeter/TBackgroundVuMeter.cpp.o CMakeFiles/libbgs.dir/package_bgs/WeightedMovingMean.cpp.o CMakeFiles/libbgs.dir/package_bgs/WeightedMovingVariance.cpp.o CMakeFiles/libbgs.dir/package_bgs/_template_/Amber.cpp.o CMakeFiles/libbgs.dir/package_bgs/_template_/MyBGS.cpp.o CMakeFiles/libbgs.dir/package_bgs/dp/AdaptiveMedianBGS.cpp.o CMakeFiles/libbgs.dir/package_bgs/dp/Eigenbackground.cpp.o CMakeFiles/libbgs.dir/package_bgs/dp/Error.cpp.o CMakeFiles/libbgs.dir/package_bgs/dp/GrimsonGMM.cpp.o CMakeFiles/libbgs.dir/package_bgs/dp/Image.cpp.o CMakeFiles/libbgs.dir/package_bgs/dp/MeanBGS.cpp.o CMakeFiles/libbgs.dir/package_bgs/dp/PratiMediodBGS.cpp.o CMakeFiles/libbgs.dir/package_bgs/dp/TextureBGS.cpp.o CMakeFiles/libbgs.dir/package_bgs/dp/WrenGA.cpp.o CMakeFiles/libbgs.dir/package_bgs/dp/ZivkovicAGMM.cpp.o CMakeFiles/libbgs.dir/package_bgs/lb/BGModel.cpp.o CMakeFiles/libbgs.dir/package_bgs/lb/BGModelFuzzyGauss.cpp.o CMakeFiles/libbgs.dir/package_bgs/lb/BGModelFuzzySom.cpp.o CMakeFiles/libbgs.dir/package_bgs/lb/BGModelGauss.cpp.o CMakeFiles/libbgs.dir/package_bgs/lb/BGModelMog.cpp.o CMakeFiles/libbgs.dir/package_bgs/lb/BGModelSom.cpp.o CMakeFiles/libbgs.dir/package_bgs/_template_/amber/amber.c.o CMakeFiles/libbgs.dir/wrapper_python/bgslibrary_module.cpp.o CMakeFiles/libbgs.dir/wrapper_python/np_opencv_converter.cpp.o CMakeFiles/libbgs.dir/wrapper_python/utils/conversion.cpp.o CMakeFiles/libbgs.dir/package_analysis/ForegroundMaskAnalysis.cpp.o CMakeFiles/libbgs.dir/package_analysis/PerformanceUtils.cpp.o CMakeFiles/libbgs.dir/package_analysis/PixelUtils.cpp.o -Wl,-rpath,/usr/local/lib -Wl,-rpath,/Users/andrewssobral/anaconda3/lib /usr/local/lib/libopencv_stitching.3.4.0.dylib /usr/local/lib/libopencv_superres.3.4.0.dylib /usr/local/lib/libopencv_videostab.3.4.0.dylib /usr/local/lib/libopencv_aruco.3.4.0.dylib /usr/local/lib/libopencv_bgsegm.3.4.0.dylib /usr/local/lib/libopencv_bioinspired.3.4.0.dylib /usr/local/lib/libopencv_ccalib.3.4.0.dylib /usr/local/lib/libopencv_dpm.3.4.0.dylib /usr/local/lib/libopencv_face.3.4.0.dylib /usr/local/lib/libopencv_fuzzy.3.4.0.dylib /usr/local/lib/libopencv_hdf.3.4.0.dylib /usr/local/lib/libopencv_img_hash.3.4.0.dylib /usr/local/lib/libopencv_line_descriptor.3.4.0.dylib /usr/local/lib/libopencv_optflow.3.4.0.dylib /usr/local/lib/libopencv_reg.3.4.0.dylib /usr/local/lib/libopencv_rgbd.3.4.0.dylib /usr/local/lib/libopencv_saliency.3.4.0.dylib /usr/local/lib/libopencv_stereo.3.4.0.dylib /usr/local/lib/libopencv_structured_light.3.4.0.dylib /usr/local/lib/libopencv_surface_matching.3.4.0.dylib /usr/local/lib/libopencv_tracking.3.4.0.dylib /usr/local/lib/libopencv_xfeatures2d.3.4.0.dylib /usr/local/lib/libopencv_ximgproc.3.4.0.dylib /usr/local/lib/libopencv_xobjdetect.3.4.0.dylib /usr/local/lib/libopencv_xphoto.3.4.0.dylib /Users/andrewssobral/anaconda3/lib/libboost_python3.dylib /usr/local/lib/libopencv_shape.3.4.0.dylib /usr/local/lib/libopencv_photo.3.4.0.dylib /usr/local/lib/libopencv_datasets.3.4.0.dylib /usr/local/lib/libopencv_plot.3.4.0.dylib /usr/local/lib/libopencv_text.3.4.0.dylib /usr/local/lib/libopencv_dnn.3.4.0.dylib /usr/local/lib/libopencv_ml.3.4.0.dylib /usr/local/lib/libopencv_video.3.4.0.dylib /usr/local/lib/libopencv_calib3d.3.4.0.dylib /usr/local/lib/libopencv_features2d.3.4.0.dylib /usr/local/lib/libopencv_highgui.3.4.0.dylib /usr/local/lib/libopencv_videoio.3.4.0.dylib /usr/local/lib/libopencv_phase_unwrapping.3.4.0.dylib /usr/local/lib/libopencv_flann.3.4.0.dylib /usr/local/lib/libopencv_imgcodecs.3.4.0.dylib /usr/local/lib/libopencv_objdetect.3.4.0.dylib /usr/local/lib/libopencv_imgproc.3.4.0.dylib /usr/local/lib/libopencv_core.3.4.0.dylib 
Undefined symbols for architecture x86_64:
  "_PyCapsule_GetPointer", referenced from:
      fs::python::init_and_export_converters() in np_opencv_converter.cpp.o
      NDArrayConverter::init() in conversion.cpp.o
  "_PyCapsule_Type", referenced from:
      fs::python::init_and_export_converters() in np_opencv_converter.cpp.o
      NDArrayConverter::init() in conversion.cpp.o
  "_PyErr_Format", referenced from:
      fs::python::init_and_export_converters() in np_opencv_converter.cpp.o
      NDArrayConverter::init() in conversion.cpp.o
  "_PyErr_Occurred", referenced from:
      expose_template_type<std::__1::vector<int, std::__1::allocator<int> > >::construct(_object*, boost::python::converter::rvalue_from_python_stage1_data*) in np_opencv_converter.cpp.o
      expose_template_type<std::__1::vector<float, std::__1::allocator<float> > >::construct(_object*, boost::python::converter::rvalue_from_python_stage1_data*) in np_opencv_converter.cpp.o
      expose_template_type<std::__1::vector<double, std::__1::allocator<double> > >::construct(_object*, boost::python::converter::rvalue_from_python_stage1_data*) in np_opencv_converter.cpp.o
      expose_template_type<std::__1::vector<cv::Point_<int>, std::__1::allocator<cv::Point_<int> > > >::construct(_object*, boost::python::converter::rvalue_from_python_stage1_data*) in np_opencv_converter.cpp.o
      expose_template_type<std::__1::vector<cv::Point_<float>, std::__1::allocator<cv::Point_<float> > > >::construct(_object*, boost::python::converter::rvalue_from_python_stage1_data*) in np_opencv_converter.cpp.o
      expose_template_type<std::__1::vector<cv::KeyPoint, std::__1::allocator<cv::KeyPoint> > >::construct(_object*, boost::python::converter::rvalue_from_python_stage1_data*) in np_opencv_converter.cpp.o
      expose_template_type<std::__1::vector<cv::Mat, std::__1::allocator<cv::Mat> > >::construct(_object*, boost::python::converter::rvalue_from_python_stage1_data*) in np_opencv_converter.cpp.o
      ...
  "_PyErr_Print", referenced from:
      fs::python::init_and_export_converters() in np_opencv_converter.cpp.o
      NDArrayConverter::init() in conversion.cpp.o
  "_PyErr_SetString", referenced from:
      fs::python::init_and_export_converters() in np_opencv_converter.cpp.o
      NDArrayConverter::init() in conversion.cpp.o
      failmsg(char const*, ...) in conversion.cpp.o
  "_PyEval_CallFunction", referenced from:
      boost::python::api::object_operators<boost::python::api::proxy<boost::python::api::attribute_policies> >::operator()() const in np_opencv_converter.cpp.o
  "_PyExc_AttributeError", referenced from:
      fs::python::init_and_export_converters() in np_opencv_converter.cpp.o
      NDArrayConverter::init() in conversion.cpp.o
  "_PyExc_ImportError", referenced from:
      fs::python::init_and_export_converters() in np_opencv_converter.cpp.o
      NDArrayConverter::init() in conversion.cpp.o
  "_PyExc_RuntimeError", referenced from:
      fs::python::init_and_export_converters() in np_opencv_converter.cpp.o
      NDArrayConverter::init() in conversion.cpp.o
  "_PyExc_TypeError", referenced from:
      failmsg(char const*, ...) in conversion.cpp.o
  "_PyFloat_FromDouble", referenced from:
      expose_template_type<std::__1::vector<float, std::__1::allocator<float> > >::convert(std::__1::vector<float, std::__1::allocator<float> > const&) in np_opencv_converter.cpp.o
      expose_template_type<std::__1::vector<double, std::__1::allocator<double> > >::convert(std::__1::vector<double, std::__1::allocator<double> > const&) in np_opencv_converter.cpp.o
      expose_template_type<std::__1::map<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, float, std::__1::less<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >, std::__1::allocator<std::__1::pair<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const, float> > > >::convert(std::__1::map<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, float, std::__1::less<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >, std::__1::allocator<std::__1::pair<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const, float> > > const&) in np_opencv_converter.cpp.o
  "_PyGILState_Ensure", referenced from:
      NumpyAllocator::allocate(int, int const*, int, void*, unsigned long*, int, cv::UMatUsageFlags) const in conversion.cpp.o
      NumpyAllocator::deallocate(cv::UMatData*) const in conversion.cpp.o
  "_PyGILState_Release", referenced from:
      NumpyAllocator::allocate(int, int const*, int, void*, unsigned long*, int, cv::UMatUsageFlags) const in conversion.cpp.o
      NumpyAllocator::deallocate(cv::UMatData*) const in conversion.cpp.o
  "_PyImport_ImportModule", referenced from:
      fs::python::init_and_export_converters() in np_opencv_converter.cpp.o
      NDArrayConverter::init() in conversion.cpp.o
  "_PyLong_FromLong", referenced from:
      expose_template_type<std::__1::vector<int, std::__1::allocator<int> > >::construct(_object*, boost::python::converter::rvalue_from_python_stage1_data*) in np_opencv_converter.cpp.o
      expose_template_type<std::__1::vector<int, std::__1::allocator<int> > >::convert(std::__1::vector<int, std::__1::allocator<int> > const&) in np_opencv_converter.cpp.o
      boost::python::api::proxy<boost::python::api::item_policies> boost::python::api::object_operators<boost::python::api::object>::operator[]<int>(int const&) in np_opencv_converter.cpp.o
      expose_template_type<std::__1::vector<float, std::__1::allocator<float> > >::construct(_object*, boost::python::converter::rvalue_from_python_stage1_data*) in np_opencv_converter.cpp.o
      expose_template_type<std::__1::vector<double, std::__1::allocator<double> > >::construct(_object*, boost::python::converter::rvalue_from_python_stage1_data*) in np_opencv_converter.cpp.o
      expose_template_type<std::__1::vector<cv::Point_<int>, std::__1::allocator<cv::Point_<int> > > >::construct(_object*, boost::python::converter::rvalue_from_python_stage1_data*) in np_opencv_converter.cpp.o
      expose_template_type<std::__1::vector<cv::Point_<float>, std::__1::allocator<cv::Point_<float> > > >::construct(_object*, boost::python::converter::rvalue_from_python_stage1_data*) in np_opencv_converter.cpp.o
      ...
  "_PyMapping_Check", referenced from:
      expose_template_type<std::__1::map<int, std::__1::vector<int, std::__1::allocator<int> >, std::__1::less<int>, std::__1::allocator<std::__1::pair<int const, std::__1::vector<int, std::__1::allocator<int> > > > > >::convertible(_object*) in np_opencv_converter.cpp.o
      expose_template_type<std::__1::map<int, std::__1::vector<float, std::__1::allocator<float> >, std::__1::less<int>, std::__1::allocator<std::__1::pair<int const, std::__1::vector<float, std::__1::allocator<float> > > > > >::convertible(_object*) in np_opencv_converter.cpp.o
      expose_template_type<std::__1::map<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, float, std::__1::less<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >, std::__1::allocator<std::__1::pair<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const, float> > > >::convertible(_object*) in np_opencv_converter.cpp.o
  "_PyObject_GetAttrString", referenced from:
      fs::python::init_and_export_converters() in np_opencv_converter.cpp.o
      NDArrayConverter::init() in conversion.cpp.o
  "_PyObject_Size", referenced from:
      expose_template_type<std::__1::vector<int, std::__1::allocator<int> > >::construct(_object*, boost::python::converter::rvalue_from_python_stage1_data*) in np_opencv_converter.cpp.o
      expose_template_type<std::__1::vector<float, std::__1::allocator<float> > >::construct(_object*, boost::python::converter::rvalue_from_python_stage1_data*) in np_opencv_converter.cpp.o
      expose_template_type<std::__1::vector<double, std::__1::allocator<double> > >::construct(_object*, boost::python::converter::rvalue_from_python_stage1_data*) in np_opencv_converter.cpp.o
      expose_template_type<std::__1::vector<cv::Point_<int>, std::__1::allocator<cv::Point_<int> > > >::construct(_object*, boost::python::converter::rvalue_from_python_stage1_data*) in np_opencv_converter.cpp.o
      expose_template_type<std::__1::vector<cv::Point_<float>, std::__1::allocator<cv::Point_<float> > > >::construct(_object*, boost::python::converter::rvalue_from_python_stage1_data*) in np_opencv_converter.cpp.o
      expose_template_type<std::__1::vector<cv::KeyPoint, std::__1::allocator<cv::KeyPoint> > >::construct(_object*, boost::python::converter::rvalue_from_python_stage1_data*) in np_opencv_converter.cpp.o
      expose_template_type<std::__1::vector<cv::Mat, std::__1::allocator<cv::Mat> > >::construct(_object*, boost::python::converter::rvalue_from_python_stage1_data*) in np_opencv_converter.cpp.o
      ...
  "_PySequence_Check", referenced from:
      expose_template_type<std::__1::vector<int, std::__1::allocator<int> > >::convertible(_object*) in np_opencv_converter.cpp.o
      expose_template_type<std::__1::vector<float, std::__1::allocator<float> > >::convertible(_object*) in np_opencv_converter.cpp.o
      expose_template_type<std::__1::vector<double, std::__1::allocator<double> > >::convertible(_object*) in np_opencv_converter.cpp.o
      expose_template_type<std::__1::vector<cv::Point_<int>, std::__1::allocator<cv::Point_<int> > > >::convertible(_object*) in np_opencv_converter.cpp.o
      expose_template_type<std::__1::vector<cv::Point_<float>, std::__1::allocator<cv::Point_<float> > > >::convertible(_object*) in np_opencv_converter.cpp.o
      expose_template_type<std::__1::vector<cv::KeyPoint, std::__1::allocator<cv::KeyPoint> > >::convertible(_object*) in np_opencv_converter.cpp.o
      expose_template_type<std::__1::vector<cv::Mat, std::__1::allocator<cv::Mat> > >::convertible(_object*) in np_opencv_converter.cpp.o
      ...
  "_PyType_IsSubtype", referenced from:
      NDArrayConverter::toMat(_object const*) in conversion.cpp.o
  "_PyUnicode_FromStringAndSize", referenced from:
      expose_template_type<std::__1::map<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, float, std::__1::less<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >, std::__1::allocator<std::__1::pair<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const, float> > > >::convert(std::__1::map<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, float, std::__1::less<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >, std::__1::allocator<std::__1::pair<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const, float> > > const&) in np_opencv_converter.cpp.o
  "_Py_Initialize", referenced from:
      fs::python::init_and_export_converters() in np_opencv_converter.cpp.o
  "__Py_NoneStruct", referenced from:
      fs::python::init_module_libbgs() in bgslibrary_module.cpp.o
      boost::python::converter::shared_ptr_from_python<bgslibrary::algorithms::FrameDifference, boost::shared_ptr>::convertible(_object*) in bgslibrary_module.cpp.o
      boost::python::converter::shared_ptr_from_python<bgslibrary::algorithms::FrameDifference, std::__1::shared_ptr>::convertible(_object*) in bgslibrary_module.cpp.o
      _object* boost::python::objects::make_instance_impl<bgslibrary::algorithms::FrameDifference, boost::python::objects::value_holder<bgslibrary::algorithms::FrameDifference>, boost::python::objects::make_instance<bgslibrary::algorithms::FrameDifference, boost::python::objects::value_holder<bgslibrary::algorithms::FrameDifference> > >::execute<boost::reference_wrapper<bgslibrary::algorithms::FrameDifference const> const>(boost::reference_wrapper<bgslibrary::algorithms::FrameDifference const> const&) in bgslibrary_module.cpp.o
      boost::python::objects::caller_py_function_impl<boost::python::detail::caller<void (*)(_object*), boost::python::default_call_policies, boost::mpl::vector2<void, _object*> > >::operator()(_object*, _object*) in bgslibrary_module.cpp.o
      boost::python::converter::shared_ptr_from_python<bgslibrary::algorithms::StaticFrameDifference, boost::shared_ptr>::convertible(_object*) in bgslibrary_module.cpp.o
      boost::python::converter::shared_ptr_from_python<bgslibrary::algorithms::StaticFrameDifference, std::__1::shared_ptr>::convertible(_object*) in bgslibrary_module.cpp.o
      ...
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make[2]: *** [libbgs.dylib] Error 1
make[1]: *** [CMakeFiles/libbgs.dir/all] Error 2
make: *** [all] Error 2
```

**Comment by Baozhang-Ren** [comment]
I built it with anaconda python3 and the newest boost and got the error I posted. 
I reinstalled boost cause with an old version it couldn't find the lib of boost even if I had specified the path to it.
Can it be the problem of boost?

**Comment by andrewssobral** [comment]
I don't know yet...
I tried this:
```
# Using conda boost
cmake -DCMAKE_VERBOSE_MAKEFILE:BOOL=ON -D BGS_PYTHON_SUPPORT=ON -D BGS_PYTHON_VERSION=3 -D PYTHON_INCLUDE_DIR=$(python -c "from distutils.sysconfig import get_python_inc; print(get_python_inc())") -D PYTHON_LIBRARY=$(python -c "import distutils.sysconfig as sysconfig; print(sysconfig.get_config_var('LIBDIR'))") -D BOOST_INCLUDEDIR=$HOME/anaconda3/include -D BOOST_LIBRARYDIR=$HOME/anaconda3/lib ..
```
and this:
```
# Using local boost
cmake -D BGS_PYTHON_SUPPORT=ON -D BGS_PYTHON_VERSION=3 -D PYTHON_INCLUDE_DIR=$(python -c "from distutils.sysconfig import get_python_inc; print(get_python_inc())") -D PYTHON_LIBRARY=$(python -c "import distutils.sysconfig as sysconfig; print(sysconfig.get_config_var('LIBDIR'))") -D BOOST_INCLUDEDIR=/Users/andrewssobral/Projects/boost_1_66_0 -D BOOST_LIBRARYDIR=/Users/andrewssobral/Projects/boost_1_66_0/stage/lib ..
```
Moreover,  I've re-compiled boost with:
```
# default
./b2 --with=all
# setting cxxflags (version 1)
./b2 --with=all cxxflags=-std=c++14
# setting cxxflags (version 2)
./b2 --with=all cxxflags="-stdlib=libstdc++" linkflags="-stdlib=libstdc++"
```
and I'm still getting: `ld: symbol(s) not found for architecture x86_64`

**Comment by andrewssobral** [comment]
By the way, this simple example works well:
```
#include <iostream>
#include <boost/python/def.hpp>
#include <boost/python/module.hpp>

using namespace std;
namespace bp = boost::python;

void hello(const char* name)
{
    cout << "Hello, " << name << "!\n";
}

BOOST_PYTHON_MODULE(mypyboost)
{
    bp::def("hello", hello);
}
```
Compiling with:
```
g++ test_python.cpp -I/Users/andrewssobral/anaconda3/include/python3.6m -I/usr/local/include/boost/python -L/Users/andrewssobral/anaconda3/lib -lboost_python3 -lpython3.6m -o mypyboost.so -fPIC -shared
```
Then:
```
python -c 'import mypyboost as my; my.hello("Andrews")'
```

**Comment by andrewssobral** [comment]
I think the best is to switch from **Boost.Python** to [**pybind11**](https://github.com/pybind/pybind11). However, I will take a while to do this...every help is welcome

**Comment by Baozhang-Ren** [comment]
Thanks! I will try that as well!

**Comment by andrewssobral** [comment]
Hi @Baozhang-Ren ,
I think the problem is the cmake version, look:
```
Boost 1.63 requires CMake 3.7 or newer
Boost 1.64 requires CMake 3.8 or newer
Boost 1.65 and 1.65.1 require CMake 3.9.3 or newer
Boost 1.66 is not yet supported by any release; Git master has support for 1.66.
```
https://stackoverflow.com/questions/42123509/cmake-finds-boost-but-the-imported-targets-not-available-for-boost-version

I will check this...

**Comment by andrewssobral** [comment]
On **Ubuntu 17.10 + Anaconda 4.4.8 + Python 3.6.3 + Boost 1.65.1**
```
conda install boost

cmake -D BGS_PYTHON_SUPPORT=ON -D BGS_PYTHON_VERSION=3 -D PYTHON_INCLUDE_DIR=$(python -c "from distutils.sysconfig import get_python_inc; print(get_python_inc())") -D PYTHON_LIBRARY=$(python -c "import distutils.sysconfig as sysconfig; print(sysconfig.get_config_var('LIBDIR'))") -D BOOST_INCLUDEDIR=$HOME/anaconda3/include -D BOOST_LIBRARYDIR=$HOME/anaconda3/lib ..
```
it works well:
```
[100%] Linking CXX shared library libbgs.so
[100%] Built target libbgs
andrews@DELL-PRECISION-M4700:~/Projects/bgslibrary/build_linx64_python3$ ls
CMakeCache.txt  CMakeFiles  cmake_install.cmake  libbgs.so  Makefile
andrews@DELL-PRECISION-M4700:~/Projects/bgslibrary/build_linx64_python3$ python
Python 3.6.3 |Anaconda custom (64-bit)| (default, Oct 13 2017, 12:02:49) 
[GCC 7.2.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import libbgs
PYTHON TYPE CONVERTERS exported
>>> quit()
```
But on Mac not yet. Maybe I need to update the CMakeFile following this:
```
#############################################################################

# This is needed so that on macs, the library is built in both 32 bit and 64 bit
# versions. Without this python might refuse to load the module, depending on
# how python was built.
# On Mac, boost needs to be compiled universal as well, if used instead of the
# included BoostParts lib. For brew, that's
# "brew install boost --universal"
# If the user chose to use the system libclang.dylib (or the libclang.dylib
# binary downloaded from llvm.org) on a mac, then we don't specify universal
# binary building since the system libclang on macs is not  universal (and thus
# linking would fail with universal).
if ( UNIVERSAL AND NOT USE_SYSTEM_LIBCLANG )
  set( CMAKE_OSX_ARCHITECTURES "i386;x86_64" )
endif()

#############################################################################

# To shut up the warning about CMake policy CMP0042
set( CMAKE_MACOSX_RPATH ON )
```
found on: https://github.com/Valloric/ycmd/blob/master/cpp/CMakeLists.txt

**Comment by Baozhang-Ren** [comment]
Thanks! I think I will do the project on Ubuntu instead.

**Comment by knaffe** [comment]
I have the same problem at Ubuntu 14.04. Actually I worked when I installed just now. But I reboot my system and  import libbgs, it rises "PYTHON TYPE CONVERTERS exported".
I update my Cmake into 3.10, use "conda install boost" to install boost.(Boost version: 1.54.0)
But it has a warning message->"WARNING: Target "libbgs" requests linking to directory "/home/chilam/anaconda3/envs/cv/lib".  Targets may link only to libraries.  CMake is dropping the item."
By the way, what is the solusion of this problem "PYTHON TYPE CONVERTERS exported"? I review the discussion but I don't know fix it at Ubuntu.
Thank you!!
